cmd/ctl/pkg/uninstall/uninstall.go

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2022 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package uninstall
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"helm.sh/helm/v3/pkg/storage/driver"
+	"log"
+	"os"
+	"time"
+
+	"github.com/cert-manager/cert-manager/cmd/ctl/pkg/build"
+	"github.com/spf13/cobra"
+	"helm.sh/helm/v3/pkg/action"
+	"helm.sh/helm/v3/pkg/cli"
+	"helm.sh/helm/v3/pkg/release"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+)
+
+type options struct {
+	settings *cli.EnvSettings
+	client   *action.Uninstall
+	cfg      *action.Configuration
+
+	DisableHooks bool
+	DryRun       bool
+	Wait         bool
+
+	genericclioptions.IOStreams
+}
+
+const (
+	defaultCertManagerNamespace = "cert-manager"
+	releaseName                 = "cert-manager"
+)
+
+func description() string {
+	return build.WithTemplate(`This command uninstalls cert-manager. It uses the Helm libraries to do so.
+
+It assumes cert-manager has been installed with this CLI.
+
+Most of the features supported by 'helm uninstall' are also supported by this command.
```

Comment:
Nit: arguably the fact that it uses Helm libraries is an implementation detail that shouldn't matter from user perspective.
It might however be good to describe what will happen if this command is run against cert-manager installation deployed by other means and also what happens with the CRDs.

Reply:
How do you suggest we delete CRDs?

https://helm.sh/docs/chart_best_practices/custom_resource_definitions

> There is no support at this time for upgrading or deleting CRDs using Helm

Reply:
> How do you suggest we delete CRDs?

At the moment if cert-manager helm release is installed with CRDs (using `--set CRDs=true` flag), then uninstalling it will also remove the CRDs.
As far as I am aware cert-manager helm release does not use Helm's CRD management mechanism due to its [caveats](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#some-caveats-and-explanations), so CRDs just end up being standard helm managed resources and uninstall works.

Perhaps it is still worth a discussion what the install/uninstall sequence should look like and I havent' thought about it in depth, but I think it should be possible to 1) run `cmctl x install [any flags]` 2) run `cmctl x uninstall [any flags]` 3) run `cmctl x install [any flags]`

Reply:
Perhaps for now this could be solved by making `cmctl x install` not fail if CRDs already exist in cluster.
The upgrade issue still remains, but I think this would make the command more user-friendly.

FWIW, I was looking at how [osm cli](https://release-v0-11.docs.openservicemesh.io/docs/guides/cli/) works for the reference- it also has install and uninstall commands, does not delete CRDs on uninstall, but also does not error out if CRDs pre-existing on install.

Reply:
> The upgrade issue still remains

Not part of this PR, but I think this will eventually need a design doc/a proper discussion what the sequence should look like and how we intend the command to be used.

Reply:
Not sure why we are discussing how the `install` command should work in this PR - I'd suggest opening an issue with the above reasoning so that we have the information in the right place and it's not lost after this is merged. That way we can continue discussions on installations flows there and focus on the changes in this PR.

On `uninstall`: 

- Removed mention of Helm libraries in the command description
- Added note for when CRDs would be deleted

Are there any other changes would you like to see here in order to approve it?

Reply:
> Not sure why we are discussing how the install command should work in this PR - I'd suggest opening an issue with the above reasoning so that we have the information in the right place and it's not lost after this is merged. That way we can continue discussions on installations flows there and focus on the changes in this PR.

I've created an issue for this https://github.com/cert-manager/cert-manager/issues/4928
I think it would be good if this got fixed in 1.8, for a better UX of `cmctl x install`/`cmctl x uninstall` in the `cmctl` 1.8 release.

Reply:
Thanks!

I'll take a look and carry on the conversation in that issue